### ISSUE TITLE
False positive after asserting instanceof grandchild

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#1208]
```php
class A {}
class B extends A {}
class C extends B {}

function bad(A $x) : void {
    if (($x instanceof C && rand(0, 1)) || rand(0, 1)) {
        return;
    }

    if ($x instanceof B) {
        if ($x instanceof C) {}
    }
}
```
Expected: No issue
Actual: RedundantCondition

https://getpsalm.org/r/942272ec4a
